Functional extensions for Result, DataResponse, DownloadResponse

groue opened the pull request:
Hello,

This pull request addresses #1072.

It contains:

- Three new Result methods: `unwrap`, `map`, and `flatMap`.
- Two new DataResponse and DownloadResponse methods: `map` and `flatMap`.
- Tests for Result
- Inline documentation

It does not contain:

- Tests for DataResponse and DownloadResponse, because I don't quite know where to put them
- Documentation in README.md, because I prefer waiting for your feedback

### Result.unwrap

`unwrap` returns a result's success value, or throws its failure error:

```swift
let result: Result<Data> = ...
try result.unwrap() // Data
```

### Result.map

`map` transforms the success value of a result, while preserving its eventual failure error. Its closure argument must not throw any error:

```swift
let result: Result<Data> = ...
result.map { $0.count } // Result<Int>
```

### Result.flatMap

`flatMap` transforms the success value of a result, while preserving its eventual failure error, and catches eventual transformation error:

```swift
let result: Result<Data> = ...
result.flatMap { try JSONSerialization.jsonObject(with: $0) } // Result<Any>
```

It is worth noting that the proposed flatMap method does not strictly follow the rules of functional programming, because its closure argument does not return another Result. Instead, its closure returns a regular value, and can throw. That's because in Swift, throwing methods are more fundamental than the `Result` type, and the sample code above looks like it is the right thing to do. Besides, users who use Result extensively can still use the unwrap method in their flatMap closures:

```swift
func f(_ data: Data) -> Result<Int> { ... }
let result: Result<Data> = ...
result.flatMap { try f($0).unwrap() } // Result<Int>
```

### DataResponse and DownloadResponse

Their `map` and `flatMap` methods apply `map` and `flatMap` to their result, while preserving all other information (requests, timelines, etc.).

They make it easier to perform ad-hoc deserializations, when extending Alamofire with a full-fledged DataResponseSerializer is not worth it.

```swift
func getStuff(_ completionHandler: @escaping (DataResponse<Stuff>) -> Void)) {
    Alamofire.request("https://httpbin.org/get").responseJSON { response in
        completionHandler(response.flatMap { JSON in
            return try Stuff(with: JSON)
        })
    }
}

getStuff { response in
    print(response.request)  // original URL request
    print(response.response) // HTTP URL response
    print(response.data)     // server data
    print(response.result)   // result of response serialization

    if let stuff = response.result.value {
        print("stuff: \(stuff)")
    }
}
```

---

What do you think?

groue commented:
cc @KingOfBrian, @cnoon, @jshier

jshier commented on Source/Result.swift:
```diff
@@ -100,3 +100,82 @@ extension Result: CustomDebugStringConvertible {
+    ///     let noString: Result<String> = .failure(error)
+    ///     try print(noString.unwrap())
+    ///     // Throws error
+    func unwrap() throws -> Value {
```
This is usually known as `dematerialize` and I think we should keep that to maintain consistency with other `Result` types (whether from Swift or other languages). At the same time, a global function called `materialize` that takes a throwing function and wraps it in a `Result` provides the opposite functionality. It would be good to have that too I think.

groue replied:
I was waiting for this question. Yes, I did not choose `dematerialize`. I don't think Alamofire needs/wants to use bleeding edge functional concepts or vocabulary.

`dematerialize` is long. `dematerialize` is spectacular. `dematerialize` is abstract. `dematerialize` is pedantic. `dematerialize` contains a latent implicit disagreement on swift's choice of try/catch/throws instead of a "proper sum type" that needs to be "materialized". `dematerialize` is not yet a term of art.

On the other side, Rust's `unwrap` sounds just as boring as it needs to be. It fits better in the multi-paradigms Swift language. It fits better with some Swift evolution discussions around wrapper types such as Optional and other various boxes. Unfortunately, this mailing list is a chore when you look for something in it, but I'd like to put my finger on a proposal/post by @erica - maybe she'll read this and help ;-)

groue replied:
Of course this is just an opinion. I'll gladly update this PR with `dematerialize` if required to do so.

groue replied:
The (unfinished) proposal by @erica is https://gist.github.com/erica/3daa8ec77aef2feaefca3a3a19aedee3, introduced in this [thread](https://lists.swift.org/pipermail/swift-evolution/Week-of-Mon-20161114/028833.html) on swift-evolution mailing list.

jshier commented:
I'm researching what `flatMap` does to the `Result` types of other languages, but for now this use looks fine.

One thing I would also like to see, even if it's not strictly part of the functional paradigm, are closures which can access the `Result`'s error without transforming them. I've defined the following extensions in my last few projects:
```swift
@discardableResult
func onSuccess(_ success: (_ value: Value) -> Void) -> Result {
    if case let .success(value) = self { success(value) }
    
    return self
}

@discardableResult
func ifSuccess(_ success: () -> Void) -> Result {
    if isSuccess { success() }
    
    return self
}

@discardableResult
func onFailure(_ failure: (_ error: Error) -> Void) -> Result {
    if case let .failure(error) = self { failure(error) }
    
    return self
}

@discardableResult
func ifFailure(_ failure: () -> Void) -> Result {
    if isFailure { failure() }
    
    return self
}
```
I think something similar could be helpful, perhaps in combination with `.mapError` and `.flatMapError` or something.

@cnoon Any input here?

groue commented:
@jshier, @cnoon: The intent behind this PR is not to turn Alamofire into a functional programming flagship, but instead to enhance Alamofire with a tiny set of functional idioms that blend well into regular imperative Swift code.

The proposal is to remove some friction in custom response deserialization. Result.map/flatMap are only there to support DataResponse.map/flatMap which are the real added value of this PR. The final touch to this PR will be *documentation* and *guidelines* for simple response deserialization. This is what this PR is about:

```swift
// Look, ma! No custom extension to DataRequest!
func getStuff(_ completionHandler: @escaping (DataResponse<Stuff>) -> Void)) {
    Alamofire.request("https://httpbin.org/get").responseJSON { response in
        completionHandler(response.flatMap { json in try Stuff(json: json) })
    }
}

getStuff { response in
    print(response.request)  // original URL request
    print(response.response) // HTTP URL response
    print(response.data)     // server data
    print(response.result)   // result of response serialization

    if let stuff = response.result.value { // stuff is Stuff
        print("stuff: \(stuff)")
    }
}
```

Alamofire *does not need* those extensions. As you say above, application developers can already add a few extensions to Result and DataResponse that cover their use cases. Alamofire is extensible, and this is good.

It just happens that custom deserialization deserves a little polish, and that a tiny set of functional idioms improves it a lot, when properly explained and documented. Embedding [antitypical/Result](https://github.com/antitypical/Result) concepts and vocabulary is not my proposal. Enhancing deserialization is.

erica commented:
> On Jan 10, 2017, at 2:56 PM, Gwendal Roué <notifications@github.com> wrote:
> 
> @groue commented on this pull request.
> 
> In Source/Result.swift <https://github.com/Alamofire/Alamofire/pull/1836>:
> 
> > @@ -100,3 +100,82 @@ extension Result: CustomDebugStringConvertible {
>          }
>      }
>  }
> +
> +// MARK: - Functional Tools
> +
> +extension Result {
> +    /// Returns the success value, or throws the failure error.
> +    ///
> +    ///     let possibleString: Result<String> = .success("success")
> +    ///     try print(possibleString.unwrap())
> +    ///     // Prints "success"
> +    ///
> +    ///     let noString: Result<String> = .failure(error)
> +    ///     try print(noString.unwrap())
> +    ///     // Throws error
> +    func unwrap() throws -> Value {
> I was waiting for this question. Yes, I did not choose dematerialize. I don't think Alamofire needs/wants to use bleeding edge functional concepts or vocabulary.
> 
> dematerialize is long. dematerialize is spectacular. dematerialize is abstract. dematerialize is pedantic. dematerialize contains a latent implicit disagreement on swift's choice of try/catch/throws instead of a "proper sum type" that needs to be "materialized". dematerialize is not yet a term of art.
> 
> On the other side, Rust's unwrap sounds just as boring as it needs to be. It fits better in the multi-paradigms Swift language. It fits better with some Swift evolution discussions around wrapper types such as Optional and other various boxes. Unfortunately, this mailing list is a chore when you look for something in it, but I'd like to put my finger on a proposal/post by @erica <https://github.com/erica> - maybe she'll read this and help ;-)
> 


You rang!? I always prefer short and sweet: `unwrap`.

-- E

jshier commented:
My main concerns with `dematerialize` vs. `unwrap` are 1) translation between other `Result` types in Swift and ours, namely antitypical/Result, which already use `dematerialize`, and 2) supporting the symmetrical operation of `materialize`, which I'm not sure works well if it's just `wrap`. But that's fine.

We want this functionality, we just have to iron out the naming and how far we want to go. Because while it may be about better response handling for you @groue, making it easier to use `Result` in any scenario is a good idea.

jshier commented:
Thinking about it more, it would probably help interop if we used `wrap`/`unwrap` and let antitypical/Result use `materialize`/`dematerialize`, so I think I've just convince myself we're good there.

erica commented:
> On Jan 11, 2017, at 11:31 AM, Jon Shier <notifications@github.com> wrote:
> 
> Thinking about it more, it would probably help interop if we used wrap/unwrap and let antitypical/Result use materialize/dematerialize, so I think I've just convince myself we're good there.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/Alamofire/Alamofire/pull/1836#issuecomment-271953638>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AAB_sMGJ00whQyj87OqnWRW9k1u4YFWcks5rRSAOgaJpZM4LJoqQ>.
> 
I think this is where I'm suppose to link a few things. 

Here you go, with the understanding that these are works in progress and likely to change.

https://gist.github.com/erica/3daa8ec77aef2feaefca3a3a19aedee3 <https://gist.github.com/erica/3daa8ec77aef2feaefca3a3a19aedee3>
https://gist.github.com/erica/db9ce92b3d23cb20799460f603c0ae7c <https://gist.github.com/erica/db9ce92b3d23cb20799460f603c0ae7c>

I have a lot of strong thoughts about `Result` types and biased unwrapping but I'm coming very late into this discussion and I'm unsure what exactly you're trying to decide.

-- E

jshier commented:
Mostly naming at this point, as we want this functionality. We're already using a `Result` type, so its existence is beyond question, so this is mainly about what other functionality it should have.

erica commented:
> On Jan 11, 2017, at 11:43 AM, Jon Shier <notifications@github.com> wrote:
> 
> Mostly naming at this point, as we want this functionality. We're already using a Result type, so its existence is beyond question, so this is mainly about what other functionality it should have.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/Alamofire/Alamofire/pull/1836#issuecomment-271956653>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AAB_sIltBU8FfBmO2pzV7_L8FpoU9OFxks5rRSK1gaJpZM4LJoqQ>.
> 
I think the unambiguous thing you are doing with a result type is unwrapping a value. You're not "materializing" it.

-- E

groue commented:
@jshier, I'l keep `unwrap` and add `wrap` when you tell me so. Although `wrap` could be bike-shedded too 😉 :

```swift
// Factory method?
Result.wrap { try value() }

// Initializer?
Result(value: { try value() })
Result(wrapping: { try value() })
```

Swift fosters initializers, isn't it? I'd vote for `Result(value: { try ... })`:

```swift
// Result<Stuff>
let result = Result(value: { try Stuff(json: json) })
result.value()      // Stuff?
try result.unwrap() // Stuff
```

cnoon commented:
Hi @groue,

My apologies for not getting back to you sooner. I wanted to make sure I had time to dig through this before giving you some direction. 😉 

I think you're right on point here. The only thing I've found that I "think" is incorrect is this code sample from your original post. Wouldn't this return `Int` or throw rather than returning a `Result<Int>` from the `unwrap`?

```swift
func f(_ data: Data) -> Result<Int> { ... }
let result: Result<Data> = ...
result.flatMap { try f($0).unwrap() } // Result<Int>
```

> Just making sure...

## Unwrap vs Dematerialize

As for unwrap vs. dematerialize...I'm much more of a fan of unwrap. Much more straightforward. Sounds like that's where all of you have landed anyways, so yay!

## Factory Method vs Initializer

As for the factory method or initializer, I'd vote initializer just like you posted in your last example.

```swift
let result = Result(value: { try Stuff(json: json) })
```

## Steps to Getting Merged

Now there are a few things we'll still need to get into this PR to get it merged.

### Result initializer

As mentioned before, let's get this added along with tests.

### Tests

Could we get tests for the `DataResponse` and `DownloadResponse`? You can just throw them into new test cases in the `Response.swift` for now.

> We're going to reorganize the test suite here at some point.

### Documentation

Let's also get the documentation added to the README. I think you need to be extra careful here though in regards to threading. You're example you posted is nice and convenient, it could be a huge bottleneck for large payloads being serialized on the main thread.

```swift
func getStuff(_ completionHandler: @escaping (DataResponse<Stuff>) -> Void)) {
    Alamofire.request("https://httpbin.org/get").responseJSON { response in
        completionHandler(response.flatMap { json in try Stuff(json: json) })
    }
}
```

While convenience is great, bad performance due to lack of understanding of what's running where is the last thing we want. Please be careful to point out these types of considerations when putting your documentation together.

Once you get the PR updated, please rebase on top of `master` (just to catch up) and we'll work to get this merged as part of the Alamofire 4.4.0 release.

Thanks again for all the hard work here and again our apologies for this being as delayed as it has been. 🍻

groue commented:
Thanks @cnoon ! For the answer, the hints, and for the warning against deserialization on the main thread, which I overlooked. Requested updates to the PR are coming :smile:

groue commented:
@cnoon, @jshier, I've rebased this PR against the master branch, added tests for DataResponse and DownloadResponse, the Result initializer reciprocal of unwrap, and presented response mapping as an introduction to custom response serialisation in the README.

I did not document the new Result methods, though! It does quite not fit in any existing section - Result is just an utility, after all. How do you think we should tell Alamofire users about these additions?

groue commented:
@cnoon I forgot your question:

> The only thing I've found that I "think" is incorrect is this code sample from your original post. Wouldn't this return Int or throw rather than returning a Result<Int> from the unwrap?
> 
> ```swift
> func f(_ data: Data) -> Result<Int> { ... }
> let result: Result<Data> = ...
> result.flatMap { try f($0).unwrap() } // Result<Int>
> ```

Yes, `f($0).unwrap()` is `Int` (or throws).

And `result.flatMap { try f($0).unwrap() }` is `Result<Int>`.

cnoon commented:
Okay @groue, I just squashed your PR down and made a few formatting and documentation tweaks and pushed it up into `master` in 7f8fe9c7 while maintaining your attribution. We can't thank you enough for all your great work here and persistence given the fact that it took us a long time to get this feature added in. I think it all paid off very well though given the final result.

These changes will ship as part of the Alamofire 4.4.0 release here shortly.

Thanks again! 🍻

erica commented:
Flicks bic light. Congrats all.

-- E

> On Feb 26, 2017, at 2:16 PM, Christian Noon <notifications@github.com> wrote:
> 
> Okay @groue <https://github.com/groue>, I just squashed your PR down and made a few formatting and documentation tweaks and pushed it up into master in 7f8fe9c <https://github.com/Alamofire/Alamofire/commit/7f8fe9c7c69a958c279827f173e3b43a90f0c552> while maintaining your attribution. We can't thank you enough for all your great work here and persistence given the fact that it took us a long time to get this feature added in. I think it all paid off very well though given the final result.
> 
> These changes will ship as part of the Alamofire 4.4.0 release here shortly.
> 
> Thanks again! 🍻
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/Alamofire/Alamofire/pull/1836#issuecomment-282588233>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AAB_sAH5hd2OVQk79zGMck3HrNJ9bj3Cks5rgeuYgaJpZM4LJoqQ>.
>

groue commented:
👍 I'm quite glad we have converged on this delicate topic! Many thanks to all, @cnoon, @jshier, @erica !!!
